crates/component-macro/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+    input: DeriveInput,
+    lifts: TokenStream,
+    loads: TokenStream,
+) -> Result<TokenStream> {
+    let internal = quote!(wasmtime::component::__internal);
+    let name = input.ident;
+    let generics = add_trait_bounds(input.generics);
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let expanded = quote! {
+        unsafe impl #impl_generics wasmtime::component::Lift for #name #ty_generics #where_clause {
+            #[inline]
+            fn lift(
+                store: &#internal::StoreOpaque,
+                options: &#internal::Options,
+                src: &Self::Lower,
+            ) -> #internal::anyhow::Result<Self> {
+                Ok(Self {
+                    #lifts
+                })
+            }
+
+            #[inline]
+            fn load(memory: &#internal::Memory, bytes: &[u8]) -> #internal::anyhow::Result<Self> {
+                debug_assert!(
+                    (bytes.as_ptr() as usize)
+                        % (<Self as wasmtime::component::ComponentType>::align() as usize)
+                        == 0
+                );
+                let mut offset = 0;
+                Ok(Self {
+                    #loads
+                })
+            }
+        }
+    };
+
+    Ok(expanded)
+}
+
+#[proc_macro_derive(Lower, attributes(component))]
+pub fn lower(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_lower(parse_macro_input!(input as DeriveInput))
+        .unwrap_or_else(Error::into_compile_error)
+        .into()
+}
+
+fn expand_lower(input: DeriveInput) -> Result<TokenStream> {
+    match find_style(&input)? {
+        Style::Record => expand_record_lower(input),
+
+        style => Err(Error::new_spanned(
+            input,
+            format!("todo: expand_lower {style:?}"),
+        )),
+    }
+}
+
+fn expand_record_lower(input: DeriveInput) -> Result<TokenStream> {
+    let name = &input.ident;
+
+    let body = if let Data::Struct(body) = &input.data {
+        body
+    } else {
+        return Err(Error::new(
+            name.span(),
+            "`record` component types can only be derived for `struct`s",
+        ));
+    };
+
+    let internal = quote!(wasmtime::component::__internal);
+
+    match &body.fields {
+        syn::Fields::Named(fields) => {
+            let mut lowers = TokenStream::new();
+            let mut stores = TokenStream::new();
+
+            for syn::Field { ty, ident, .. } in &fields.named {
+                lowers.extend(quote!(wasmtime::component::Lower::lower(
+                    &self.#ident, store, options, unsafe {
+                        // Extract a mutable pointer to the field and transmute it to a `&mut MaybeUninit<#ty>`,
+                        // taking care to never create a `&mut Self::Lower` or a `&mut #ty`, which would be
+                        // unsound.
+                        std::mem::transmute::<_, _>(std::ptr::addr_of_mut!((*dst.as_mut_ptr()).#ident))
```

Comment:
I'm not actually sure if this is possible, but if we can I think it'd be best to reuse the `map_maybe_uninit!` macro used internally here as well. If that doesn't work out though can you fill in the two type parameters to `transmute` here to clarify what's happening?

Reply:
Yeah, I started out reusing `map_maybe_uninit!`, but that required adding a `#[macro_export]` and making `MaybeUninitExt` public.  If that sounds okay to you, I'll do that.

Reply:
That should be fine as long as they're all marked `#[doc(hidden)]` to not actually show up in the public documentation